### ISSUE TITLE
Blockdevice duplication upon device path change

### ISSUE THREAD

**Issue by nike38rus** [openebs/openebs#2823]
Hey there!

There are some cases when block device device path can change. E.g. if you have 3 disks: /dev/sda, /dev/sdb and /dev/sdc and you need to remove second one. You just detach /dev/sdb and then you have /dev/sda and /dev/sdc disks. After server reboot /dev/sdc will become /dev/sdb. In such situations OpenEBS NDM discovers new disk with new device path and creates corresponding BD resource, marks it as Active and marks previous disk and BD with old devicepath as inactive.

After those changes system seems working properly, although storagePoolClaim contains outdated BD list and new BD does not have BD claim. New and old disks and BDs differs only with spec.path and metadata.name fields. 

The same situation occurs when there is another storage provider in the cluster and block devices can be detached from one cluster node and attached to another. I use vSphere cloud provider in parallel with OpenEBS and I can't exclude disks by vendor name because I use the same vSphere disks in OpenEBS  Every block device reattach cause BD duplication.

It seems as NDM discovery routine misbehaving and it may cause problems during storage pool maintenance (diskreplacement, pool extension etc.). In general, block device paths and their attach order may vary. This should be considered in order to guarantee system stability and predictablilty.

Steps to reproduce:
1. Add new k8s node into OpenEBS-managed cluster with one disk (/dev/sda)
2. Attach a disk to it. It will show up as /dev/sdb
3. Detach /dev/sdb disk from node
4. Attach new disk to node. It will show up as /dev/sdc
5. Reboot node. Disk /dev/sdc will be renamed to /dev/sdb. OpenEBS will discover new disk and create new BD resource. All specs except spec.path are the same

```
$ k describe disk disk-a326fcbd38e0e566545abe08da964894 
Name:         disk-a326fcbd38e0e566545abe08da964894
Namespace:    
Labels:       kubernetes.io/hostname=k8s-node02-lab
              ndm.io/disk-type=disk
              ndm.io/managed=true
Annotations:  <none>
API Version:  openebs.io/v1alpha1
Device Info:
  Block Device UID:  
Kind:                Disk
Metadata:
  Creation Timestamp:  2019-10-24T08:14:06Z
  Generation:          6
  Resource Version:    34076076
  Self Link:           /apis/openebs.io/v1alpha1/disks/disk-a326fcbd38e0e566545abe08da964894
  UID:                 3f8b88e3-f636-11e9-a7e3-00505683d5d2
Spec:
  Capacity:
    Logical Sector Size:   512
    Physical Sector Size:  512
    Storage:               53687091200
  Details:
    Compliance:         SPC-4
    Drive Type:         Not Available
    Firmware Revision:  2.0 
    Model:              Virtual_disk
    Rotation Rate:      0
    Serial:             6000c2934e489a8eff32cf47d1e108ad
    Vendor:             VMware
  Devlinks:
    Kind:  by-id
    Links:
      /dev/disk/by-id/scsi-36000c2934e489a8eff32cf47d1e108ad
      /dev/disk/by-id/wwn-0x6000c2934e489a8eff32cf47d1e108ad
    Kind:  by-path
    Links:
      /dev/disk/by-path/fc---lun-0
      /dev/disk/by-path/pci-0000:03:00.0-scsi-0:0:13:0
  File System:
  Path:  /dev/sdm
Stats:
  Device Utilization Rate:  0
  Disk Temperature:
    Current Temperature:   0
    Highest Temperature:   0
    Lowest Temperature:    0
  Percent Endurance Used:  0
  Total Bytes Read:        0
  Total Bytes Written:     0
Status:
  State:  Inactive
Events:   <none>
```
```
$ k describe disk disk-3a4c593577c134bf9dc993acf882413e
Name:         disk-3a4c593577c134bf9dc993acf882413e
Namespace:    
Labels:       kubernetes.io/hostname=k8s-node02-lab
              ndm.io/disk-type=disk
              ndm.io/managed=true
Annotations:  <none>
API Version:  openebs.io/v1alpha1
Device Info:
  Block Device UID:  
Kind:                Disk
Metadata:
  Creation Timestamp:  2019-10-28T03:48:54Z
  Generation:          1
  Resource Version:    34076079
  Self Link:           /apis/openebs.io/v1alpha1/disks/disk-3a4c593577c134bf9dc993acf882413e
  UID:                 c95f4562-e304-44e2-8f2b-f482f7a3444d
Spec:
  Capacity:
    Logical Sector Size:   512
    Physical Sector Size:  512
    Storage:               53687091200
  Details:
    Compliance:         SPC-4
    Drive Type:         Not Available
    Firmware Revision:  2.0 
    Model:              Virtual_disk
    Rotation Rate:      0
    Serial:             6000c2934e489a8eff32cf47d1e108ad
    Vendor:             VMware
  Devlinks:
    Kind:  by-id
    Links:
      /dev/disk/by-id/scsi-36000c2934e489a8eff32cf47d1e108ad
      /dev/disk/by-id/wwn-0x6000c2934e489a8eff32cf47d1e108ad
    Kind:  by-path
    Links:
      /dev/disk/by-path/pci-0000:03:00.0-scsi-0:0:13:0
      /dev/disk/by-path/fc---lun-0
  File System:
  Path:  /dev/sdb
Stats:
  Device Utilization Rate:  0
  Disk Temperature:
    Current Temperature:   0
    Highest Temperature:   0
    Lowest Temperature:    0
  Percent Endurance Used:  0
  Total Bytes Read:        0
  Total Bytes Written:     0
Status:
  State:  Active
Events:   <none>
```

**Comment by akhilerm** [comment]
@nike38rus  Apologies for the late reply. Since you are using vsphere, can you check if the steps mentioned [here](https://blog.openebs.io/uniquely-identifying-disks-in-openebs-on-vmware-platform-a69912ab99fe), help you to uniquely identify the disk.

If the UUID is enabled as per the blog, I think the problem which occurs due to path change should be resolved.

Also to fix the the NDM discovery routine for virtual disks, the changes are currently being reviewed under [this PR](https://github.com/openebs/node-disk-manager/pull/386).

**Comment by nike38rus** [comment]
@akhilerm that's already done due to vSphere storage provider requirements. And it does not help.

**Comment by akhilerm** [comment]
@nike38rus The changes for uniquely identifying disks has been merged, It can be enabled via a feature flag. More details can be found here. [ReleaseNotes](https://github.com/openebs/openebs/wiki/Change-Summary-v1.10#alpha-feature-updates)->[NDM]

**Comment by akhilerm** [comment]
The unique identification of virtual disks is available as a default feature starting from OpenEBS 2.0. More details are available  [here](https://github.com/openebs/openebs/wiki/Change-Summary-v2.0#new-capabilities).
